chain/src/Pos/Chain/Ssc/Mem.hs

```diff
@@ -80,21 +88,26 @@ sscRunLocalQuery action = do
 -- | Run STM transaction which modifies 'SscLocalData' and also can log.
 sscRunLocalSTM
     :: forall ctx m a.
-       (MonadSscMem ctx m, MonadIO m, WithLogger m)
-    => StateT SscLocalData (NamedPureLogger STM) a -> m a
-sscRunLocalSTM action = do
+       (MonadSscMem ctx m, MonadIO m)
+    => TraceNamed m
+    -> WriterT (DList LoggedItem) (StateT SscLocalData STM) a
+    -> m a
+sscRunLocalSTM logTrace action = do
     localVar <- sscLocal <$> askSscMem
-    launchNamedPureLog atomically $ syncingStateWith localVar action
+    (a, logItems) <- atomically $ syncingStateWith localVar $ runWriterT action
+    forM_ (DList.toList logItems) (traceWith logTrace)
+    pure a
 
 ----------------------------------------------------------------------------
 -- Global
 ----------------------------------------------------------------------------
 
-type SscGlobalQuery a =
-    forall m . (MonadReader SscGlobalState m) => m a
+type SscGlobalQuery a = forall m . Monad m =>
+    TraceNamed m ->
+    ReaderT SscGlobalState m a
 
 type SscGlobalUpdate a =
-    forall m . (MonadState SscGlobalState m, WithLogger m, Rand.MonadRandom m) => m a
+    WriterT (DList LoggedItem) (StateT SscGlobalState (Rand.MonadPseudoRandom Rand.ChaChaDRG)) a
```

Comment:
I'm very skeptical of `WriterT` due to it's propensity for space leaks (both in thunk accumulation and lack-of-streaming). It would be preferable to use `StateT` to avoid thunk accumulation.